src/TurbulenceClosures/turbulence_closure_implementations/TKEBasedVerticalDiffusivities/TKEBasedVerticalDiffusivities.jl

```diff
@@ -0,0 +1,186 @@
+module TKEBasedVerticalDiffusivities
+
+using Adapt
+using CUDA
+using KernelAbstractions: @kernel, @index
+
+using Oceananigans.Architectures
+using Oceananigans.Grids
+using Oceananigans.Utils
+using Oceananigans.Units
+using Oceananigans.Fields
+using Oceananigans.Operators
+
+using Oceananigans.Utils: prettysummary
+using Oceananigans.Grids: peripheral_node, inactive_node, inactive_cell
+using Oceananigans.Fields: ZeroField
+using Oceananigans.BoundaryConditions: default_prognostic_bc, DefaultBoundaryCondition
+using Oceananigans.BoundaryConditions: BoundaryCondition, FieldBoundaryConditions
+using Oceananigans.BoundaryConditions: DiscreteBoundaryFunction, FluxBoundaryCondition
+using Oceananigans.BuoyancyModels: ∂z_b, top_buoyancy_flux
+using Oceananigans.Grids: inactive_cell
+
+using Oceananigans.TurbulenceClosures:
+    getclosure,
+    time_discretization,
+    AbstractScalarDiffusivity,
+    VerticallyImplicitTimeDiscretization,
+    VerticalFormulation
+    
+import Oceananigans.BoundaryConditions: getbc
+import Oceananigans.Utils: with_tracers
+import Oceananigans.TurbulenceClosures:
+    validate_closure,
+    shear_production,
+    buoyancy_flux,
+    dissipation,
+    add_closure_specific_boundary_conditions,
+    compute_diffusivities!,
+    DiffusivityFields,
+    implicit_linear_coefficient,
+    viscosity,
+    diffusivity,
+    viscosity_location,
+    diffusivity_location,
+    diffusive_flux_x,
+    diffusive_flux_y,
+    diffusive_flux_z
+
+const c = Center()
+const f = Face()
+const VITD = VerticallyImplicitTimeDiscretization
+
+@inline ϕ²(i, j, k, grid, ϕ, args...) = ϕ(i, j, k, grid, args...)^2
+
+@inline function shearᶜᶜᶠ(i, j, k, grid, u, v)
+    ∂z_u² = ℑxᶜᵃᵃ(i, j, k, grid, ϕ², ∂zᶠᶜᶠ, u)
+    ∂z_v² = ℑyᵃᶜᵃ(i, j, k, grid, ϕ², ∂zᶜᶠᶠ, v)
+    S² = ∂z_u² + ∂z_v²
+    return S²
+end
+
+@inline function shearᶜᶜᶜ(i, j, k, grid, u, v)
+    ∂z_u² = ℑxᶜᵃᵃ(i, j, k, grid, ℑbzᵃᵃᶜ, ϕ², ∂zᶠᶜᶠ, u)
+    ∂z_v² = ℑyᵃᶜᵃ(i, j, k, grid, ℑbzᵃᵃᶜ, ϕ², ∂zᶜᶠᶠ, v)
+    S² = ∂z_u² + ∂z_v²
+    return S²
+end
+
+@inline Riᶜᶜᶜ(i, j, k, grid, velocities, tracers, buoyancy) =
+    ℑbzᵃᵃᶜ(i, j, k, grid, Riᶜᶜᶠ, velocities, tracers, buoyancy)
+
+@inline function Riᶜᶜᶠ(i, j, k, grid, velocities, tracers, buoyancy)
+    ∂z_u² = ℑxᶜᵃᵃ(i, j, k, grid, ϕ², ∂zᶠᶜᶠ, velocities.u)
+    ∂z_v² = ℑyᵃᶜᵃ(i, j, k, grid, ϕ², ∂zᶜᶠᶠ, velocities.v)
+    N² = ∂z_b(i, j, k, grid, buoyancy, tracers)
+    S² = ∂z_u² + ∂z_v²
```

Comment:
isn't this repetitive? why not use `shearᶜᶜᶠ` here?

Reply:
I think we could yes